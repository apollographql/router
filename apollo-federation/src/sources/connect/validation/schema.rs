//! Validations that check the entire connectors schema together:

use apollo_compiler::Name;
use apollo_compiler::Schema;
use apollo_compiler::ast::Directive;
use apollo_compiler::collections::IndexSet;
use apollo_compiler::executable::FieldSet;
use apollo_compiler::executable::Selection;
use apollo_compiler::name;
use apollo_compiler::parser::Parser;
use apollo_compiler::parser::SourceMap;
use apollo_compiler::parser::SourceSpan;
use apollo_compiler::schema::Component;
use apollo_compiler::schema::ExtendedType;
use apollo_compiler::schema::ObjectType;
use apollo_compiler::validation::Valid;
use hashbrown::HashSet;
use indexmap::IndexMap;
use itertools::Itertools;
use shape::Shape;
use shape::ShapeCase;
use shape::ShapeVisitor;

use self::keys::EntityKeyChecker;
use self::keys::field_set_error;
use crate::link::Import;
use crate::link::Link;
use crate::link::federation_spec_definition::FEDERATION_FIELDS_ARGUMENT_NAME;
use crate::link::federation_spec_definition::FEDERATION_KEY_DIRECTIVE_NAME_IN_SPEC;
use crate::link::federation_spec_definition::FEDERATION_RESOLVABLE_ARGUMENT_NAME;
use crate::link::spec::Identity;
use crate::sources::connect::Connector;
use crate::sources::connect::json_selection::SelectionTrie;
use crate::sources::connect::validation::Code;
use crate::sources::connect::validation::Message;
use crate::sources::connect::validation::graphql::SchemaInfo;
use crate::subgraph::spec::CONTEXT_DIRECTIVE_NAME;
use crate::subgraph::spec::EXTERNAL_DIRECTIVE_NAME;
use crate::subgraph::spec::FROM_CONTEXT_DIRECTIVE_NAME;

mod keys;

pub(super) fn validate(
    schema: &SchemaInfo,
    file_name: &str,
    fields_seen_by_connectors: Vec<(Name, Name)>,
) -> Vec<Message> {
    let messages: Vec<Message> = check_for_disallowed_type_definitions(schema)
        .chain(check_conflicting_directives(schema))
        .collect();
    if !messages.is_empty() {
        return messages;
    }
    check_seen_fields(schema, fields_seen_by_connectors)
        .chain(advanced_validations(schema, file_name))
        .collect()
}

fn check_for_disallowed_type_definitions(schema: &SchemaInfo) -> impl Iterator<Item = Message> {
    let subscription_name = schema
        .schema_definition
        .subscription
        .as_ref()
        .map(|sub| &sub.name);
    schema
        .types
        .values()
        .filter_map(move |extended_type| match extended_type {
            ExtendedType::Union(union_type) => Some(abstract_type_error(
                SourceSpan::recompose(union_type.location(), union_type.name.location()),
                &schema.sources,
                "union",
            )),
            ExtendedType::Interface(interface) => Some(abstract_type_error(
                SourceSpan::recompose(interface.location(), interface.name.location()),
                &schema.sources,
                "interface",
            )),
            ExtendedType::Object(obj) if subscription_name.is_some_and(|name| name == &obj.name) => {
                    Some(Message {
                        code: Code::SubscriptionInConnectors,
                        message: format!(
                            "A subscription root type is not supported when using `@{connect_directive_name}`.",
                            connect_directive_name = schema.connect_directive_name(),
                        ),
                        locations: obj.name.line_column_range(&schema.sources).into_iter().collect(),
                    })
            }
            _ => None,
        })
}

/// Certain federation directives are not allowed when using connectors.
/// We produce errors for any which were imported, even if not used.
fn check_conflicting_directives(schema: &Schema) -> Vec<Message> {
    let Some((fed_link, fed_link_directive)) =
        Link::for_identity(schema, &Identity::federation_identity())
    else {
        return Vec::new();
    };

    // TODO: make the `Link` code retain locations directly instead of reparsing stuff for validation
    let imports = fed_link_directive
        .specified_argument_by_name(&name!("import"))
        .and_then(|arg| arg.as_list())
        .into_iter()
        .flatten()
        .filter_map(|value| Import::from_value(value).ok().map(|import| (value, import)))
        .collect_vec();

    let disallowed_imports = [CONTEXT_DIRECTIVE_NAME, FROM_CONTEXT_DIRECTIVE_NAME];
    fed_link
        .imports
        .into_iter()
        .filter_map(|import| {
            disallowed_imports
                .contains(&import.element)
                .then(|| Message {
                    code: Code::ConnectorsUnsupportedFederationDirective,
                    message: format!(
                        "The directive `@{import}` is not supported when using connectors.",
                        import = import.alias.as_ref().unwrap_or(&import.element)
                    ),
                    locations: imports
                        .iter()
                        .find_map(|(value, reparsed)| {
                            (*reparsed == *import).then(|| value.line_column_range(&schema.sources))
                        })
                        .flatten()
                        .into_iter()
                        .collect(),
                })
        })
        .collect()
}

fn abstract_type_error(node: Option<SourceSpan>, source_map: &SourceMap, keyword: &str) -> Message {
    Message {
        code: Code::ConnectorsUnsupportedAbstractType,
        message: format!(
            "Abstract schema types, such as `{keyword}`, are not supported when using connectors. You can check out our documentation at https://go.apollo.dev/connectors/best-practices#abstract-schema-types-are-unsupported."
        ),
        locations: node
            .and_then(|location| location.line_column_range(source_map))
            .into_iter()
            .collect(),
    }
}

/// Check that all fields defined in the schema are resolved by a connector.
fn check_seen_fields(
    schema: &SchemaInfo,
    fields_seen_by_connectors: Vec<(Name, Name)>,
) -> impl Iterator<Item = Message> {
    let federation = Link::for_identity(schema, &Identity::federation_identity());
    let external_directive_name = federation.map_or(EXTERNAL_DIRECTIVE_NAME, |(link, _)| {
        link.directive_name_in_schema(&EXTERNAL_DIRECTIVE_NAME)
    });

    let all_fields: IndexSet<_> = schema
        .types
        .values()
        .filter_map(|extended_type| {
            if extended_type.is_built_in() {
                return None;
            }
            let coord = |(name, _): (&Name, _)| (extended_type.name().clone(), name.clone());

            // ignore all fields on objects marked @external
            if extended_type
                .directives()
                .iter()
                .any(|dir| dir.name == external_directive_name)
            {
                return None;
            }

            match extended_type {
                ExtendedType::Object(object) => {
                    // ignore fields marked @external
                    Some(
                        object
                            .fields
                            .iter()
                            .filter(|(_, def)| {
                                !def.directives
                                    .iter()
                                    .any(|dir| dir.name == external_directive_name)
                            })
                            .map(coord),
                    )
                }
                ExtendedType::Interface(_) => None, // TODO: when interfaces are supported (probably should include fields from implementing/member types as well)
                _ => None,
            }
        })
        .flatten()
        .collect();

    let mut seen_fields = fields_seen_by_resolvable_keys(schema);
    seen_fields.extend(fields_seen_by_connectors);

    (&all_fields - &seen_fields).into_iter().map(move |(parent_type, field_name)| {
        let Ok(field_def) = schema.type_field(&parent_type, &field_name) else {
            // This should never happen, but if it does, we don't want to panic
            return Message {
                code: Code::GraphQLError,
                message: format!(
                    "Field `{parent_type}.{field_name}` is missing from the schema.",
                ),
                locations: Vec::new(),
            };
        };
        Message {
            code: Code::ConnectorsUnresolvedField,
            message: format!(
                "No connector resolves field `{parent_type}.{field_name}`. It must have a `@{connect_directive_name}` directive or appear in `@{connect_directive_name}(selection:)`.",
                connect_directive_name = schema.connect_directive_name()
            ),
            locations: field_def.line_column_range(&schema.sources).into_iter().collect(),
        }
    })
}

fn fields_seen_by_resolvable_keys(schema: &SchemaInfo) -> IndexSet<(Name, Name)> {
    let mut seen_fields = IndexSet::default();
    let objects = schema.types.values().filter_map(|node| node.as_object());
    // Mark resolvable key fields as seen
    let mut selections: Vec<(Name, Selection)> = objects
        .clone()
        .flat_map(|object| {
            resolvable_key_fields(object, schema).flat_map(|(field_set, _)| {
                field_set
                    .selection_set
                    .selections
                    .iter()
                    .map(|selection| (object.name.clone(), selection.clone()))
                    .collect::<Vec<_>>()
            })
        })
        .collect();
    while !selections.is_empty() {
        if let Some((type_name, selection)) = selections.pop() {
            if let Some(field) = selection.as_field() {
                let t = (type_name, field.name.clone());
                if !seen_fields.contains(&t) {
                    seen_fields.insert(t);
                    field.selection_set.selections.iter().for_each(|selection| {
                        selections.push((field.ty().inner_named_type().clone(), selection.clone()));
                    });
                }
            }
        }
    }

    seen_fields
}

/// For an object type, get all the keys (and directive nodes) that are resolvable.
///
/// The [`FieldSet`] returned here is what goes in the `fields` argument, so `id` in `@key(fields: "id")`
fn resolvable_key_fields<'a>(
    object: &'a ObjectType,
    schema: &'a Schema,
) -> impl Iterator<Item = (FieldSet, &'a Component<Directive>)> {
    object
        .directives
        .iter()
        .filter(|directive| directive.name == FEDERATION_KEY_DIRECTIVE_NAME_IN_SPEC)
        .filter(|directive| {
            directive
                .arguments
                .iter()
                .find(|arg| arg.name == FEDERATION_RESOLVABLE_ARGUMENT_NAME)
                .and_then(|arg| arg.value.to_bool())
                .unwrap_or(true)
        })
        .filter_map(|directive| {
            directive
                .arguments
                .iter()
                .find(|arg| arg.name == FEDERATION_FIELDS_ARGUMENT_NAME)
                .map(|arg| &arg.value)
                .and_then(|value| value.as_str())
                .and_then(|fields_str| {
                    Parser::new()
                        .parse_field_set(
                            Valid::assume_valid_ref(schema),
                            object.name.clone(),
                            fields_str.to_string(),
                            "",
                        )
                        .ok()
                        .map(|field_set| (field_set, directive))
                })
        })
}

fn advanced_validations(schema: &SchemaInfo, subgraph_name: &str) -> Vec<Message> {
    let mut messages = Vec::new();

    let Ok(connectors) = Connector::from_schema(schema, subgraph_name, schema.connect_link.spec())
    else {
        return messages;
    };

    let mut entity_checker = EntityKeyChecker::default();

    for (field_set, directive) in find_all_resolvable_keys(schema) {
        entity_checker.add_key(&field_set, directive);
    }

    for (_, connector) in &connectors {
        if connector.entity_resolver == Some(crate::sources::connect::EntityResolver::TypeBatch) {
            let input_trie = compute_input_trie(connector);
            match SelectionSetWalker::new(&input_trie).walk(&connector.selection.shape(), connector)
            {
                Ok(res) => messages.extend(res),
                Err(err) => messages.push(err),
            }
        }
    }

    for (_, connector) in connectors {
        match connector.resolvable_key(schema) {
            Ok(None) => continue,
            Err(_) => {
                let variables = connector.variable_references().collect_vec();
                messages.push(field_set_error(
                    &variables,
                    &connector.id.directive.coordinate(),
                ));
            }
            Ok(Some(field_set)) => {
                entity_checker.add_connector(field_set);
            }
        }
    }

    if !messages.is_empty() {
        // Don't produce errors about unresolved keys if we _know_ some of the generated keys are wrong
        return messages;
    }

    entity_checker.check_for_missing_entity_connectors(schema)
}

fn compute_input_trie(connector: &Connector) -> SelectionTrie {
    let mut merged = SelectionTrie::new();
    for param in connector.variable_references() {
        merged.extend(&param.selection);
    }
    merged
}

struct SelectionSetWalker<'walker> {
    trie: &'walker SelectionTrie,
    unmapped_fields: HashSet<String>,
}

impl<'walker> SelectionSetWalker<'walker> {
    fn new(trie: &'walker SelectionTrie) -> Self {
        SelectionSetWalker {
            trie,
            unmapped_fields: HashSet::new(),
        }
    }
}

#[derive(Debug, thiserror::Error)]
enum ShapeVisitorError<'error> {
    #[error(
        "The `@connect` directive on `{connector}` specifies a batch entity resolver, but the key `{unset}` is not a subset of the output shape `{output_shape}`"
    )]
    BatchKeyNotSubsetOfOutputShape {
        connector: String,
        unset: &'error String,
        output_shape: &'error Shape,
    },
    #[error("Attempted to resolve key on unexpected shape `{shape_str}`")]
    UnexpectedKeyOnShape { shape_str: String },
    #[error("Batch keys cannot be mapped from non-`$root` sources (i.e. `$context`, `$this`)")]
    NonRootBatch,
}

#[allow(clippy::extra_unused_lifetimes)]
impl<'error> From<ShapeVisitorError<'_>> for Message {
    fn from(value: ShapeVisitorError) -> Self {
        match &value {
            ShapeVisitorError::BatchKeyNotSubsetOfOutputShape { .. } => Message {
                code: Code::ConnectorsUnresolvedField,
                message: value.to_string(),
                locations: Vec::new(),
            },
            ShapeVisitorError::UnexpectedKeyOnShape { .. } => Message {
                code: Code::ConnectorsUnresolvedField,
                message: value.to_string(),
                locations: Vec::new(),
            },
            ShapeVisitorError::NonRootBatch => Message {
                code: Code::ConnectBatchAndThis,
                message: value.to_string(),
                locations: Vec::new(),
            },
        }
    }
}

impl SelectionSetWalker<'_> {
    fn walk(
        mut self,
        output_shape: &Shape,
        connector: &Connector,
    ) -> Result<Vec<Message>, Message> {
        output_shape.visit_shape(&mut self)?;

        // Collect messages from unset Names
        let mut vec = Vec::new();
        for unset in &self.unmapped_fields {
            vec.push(
                ShapeVisitorError::BatchKeyNotSubsetOfOutputShape {
                    connector: connector.id.directive.synthetic_name(),
                    unset,
                    output_shape,
                }
                .into(),
            );
        }
        Ok(vec)
    }
}
impl<'walker> ShapeVisitor for SelectionSetWalker<'walker> {
    type Error = ShapeVisitorError<'walker>;
    type Output = ();

    fn default(&mut self, shape: &Shape) -> Result<Self::Output, Self::Error> {
        Err(ShapeVisitorError::UnexpectedKeyOnShape {
            shape_str: shape.pretty_print(),
        })
    }

    fn visit_object(
        &mut self,
        _: &Shape,
        fields: &IndexMap<String, Shape>,
        _: &Shape,
    ) -> Result<Self::Output, Self::Error> {
        for (key, sub_selection) in self.trie.iter() {
            // Object should contain all keys in the selection set.
            // If not, then the key is unmapped.
            if !fields.contains_key(key) {
                self.unmapped_fields.insert(key.to_string());
                continue;
            }

            let Some(next_shape) = fields.get(key) else {
                unreachable!("nested shape is expected")
            };

            // Check that next shape doesn't come from a non-`$root` field.
            if let ShapeCase::Name(root, _) = next_shape.case() {
                if root.value != "$root" {
                    return Err(ShapeVisitorError::NonRootBatch);
                }
            }

            // If key has no nested selections, then we can stop walking down this branch.
            if sub_selection.is_empty() {
                continue;
            }

            // Continue walking with nested selection sets
            let mut nested = SelectionSetWalker::new(sub_selection);
            next_shape.visit_shape(&mut nested)?;
            self.unmapped_fields
                .extend(nested.unmapped_fields.into_iter());
        }
        Ok(())
    }
}

fn find_all_resolvable_keys(schema: &Schema) -> Vec<(FieldSet, &Component<Directive>)> {
    schema
        .types
        .values()
        .filter_map(|extended_type| extended_type.as_object())
        .flat_map(|object| resolvable_key_fields(object, schema))
        .collect()
}
