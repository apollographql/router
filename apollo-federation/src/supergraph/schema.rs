use apollo_compiler::schema::SchemaBuilder;

use crate::error::FederationError;
use crate::schema::FederationSchema;

/// TODO: Use the JS/programmatic approach instead of hard-coding definitions.
pub(crate) fn new_empty_fed_2_subgraph_schema() -> Result<FederationSchema, FederationError> {
    let builder = SchemaBuilder::new().adopt_orphan_extensions();
    let builder = builder.parse(
        r#"
    extend schema
        @link(url: "https://specs.apollo.dev/link/v1.0")
        @link(url: "https://specs.apollo.dev/federation/v2.9")

    directive @link(url: String, as: String, for: link__Purpose, import: [link__Import]) repeatable on SCHEMA

    scalar link__Import

    enum link__Purpose {
        """
        \`SECURITY\` features provide metadata necessary to securely resolve fields.
        """
        SECURITY

        """
        \`EXECUTION\` features provide metadata necessary for operation execution.
        """
        EXECUTION
    }

    directive @federation__key(fields: federation__FieldSet!, resolvable: Boolean = true) repeatable on OBJECT | INTERFACE

    directive @federation__requires(fields: federation__FieldSet!) on FIELD_DEFINITION

    directive @federation__provides(fields: federation__FieldSet!) on FIELD_DEFINITION

    directive @federation__external(reason: String) on OBJECT | FIELD_DEFINITION

    directive @federation__tag(name: String!) repeatable on FIELD_DEFINITION | OBJECT | INTERFACE | UNION | ARGUMENT_DEFINITION | SCALAR | ENUM | ENUM_VALUE | INPUT_OBJECT | INPUT_FIELD_DEFINITION | SCHEMA

    directive @federation__extends on OBJECT | INTERFACE

    directive @federation__shareable on OBJECT | FIELD_DEFINITION

    directive @federation__inaccessible on FIELD_DEFINITION | OBJECT | INTERFACE | UNION | ARGUMENT_DEFINITION | SCALAR | ENUM | ENUM_VALUE | INPUT_OBJECT | INPUT_FIELD_DEFINITION

    directive @federation__override(from: String!, label: String) on FIELD_DEFINITION

    directive @federation__composeDirective(name: String) repeatable on SCHEMA

    directive @federation__interfaceObject on OBJECT

    directive @federation__authenticated on FIELD_DEFINITION | OBJECT | INTERFACE | SCALAR | ENUM

    directive @federation__requiresScopes(scopes: [[federation__Scope!]!]!) on FIELD_DEFINITION | OBJECT | INTERFACE | SCALAR | ENUM

    directive @federation__cost(weight: Int!) on ARGUMENT_DEFINITION | ENUM | FIELD_DEFINITION | INPUT_FIELD_DEFINITION | OBJECT | SCALAR

    directive @federation__listSize(assumedSize: Int, slicingArguments: [String!], sizedFields: [String!], requireOneSlicingArgument: Boolean = true) on FIELD_DEFINITION

    directive @federation__fromContext(field: federation__ContextFieldValue) on ARGUMENT_DEFINITION

    directive @federation__context(name: String!) repeatable on INTERFACE | OBJECT | UNION

    scalar federation__ContextFieldValue

    scalar federation__FieldSet

    scalar federation__Scope
    "#,
        "subgraph.graphql",
    );
    FederationSchema::new(builder.build()?)
}
