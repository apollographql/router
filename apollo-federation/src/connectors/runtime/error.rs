use serde_json_bytes::{ByteString, Map, Value};

#[derive(Debug, Default)]
pub struct RawError {
    pub message: String,
    pub code: Option<ByteString>,
    pub extensions: Map<ByteString, Value>,
    pub path: Option<ByteString>,
}

impl RawError {
    pub fn extension_code(mut self, code: impl Into<String>) -> Self {
        self.code = Some(code.into().into());
        self
    }

    pub fn extension<K, V>(mut self, key: K, value: V) -> Self
    where
        K: Into<ByteString>,
        V: Into<Value>,
    {
        self.extensions.insert(key.into(), value.into());
        self
    }
}
