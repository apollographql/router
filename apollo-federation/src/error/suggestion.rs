use levenshtein::levenshtein;

use crate::utils::human_readable;

pub(crate) fn suggestion_list(
    input: &str,
    options: impl IntoIterator<Item = String>,
) -> Vec<String> {
    let threshold = 1 + (input.len() as f64 * 0.4).floor() as usize;
    let input_lowercase = input.to_lowercase();
    let mut result = Vec::new();
    for option in options {
        // Special casing so that if the only mismatch is in upper/lower-case, then the option is
        // always shown.
        let distance = if input_lowercase == option.to_lowercase() {
            1
        } else {
            levenshtein(input, &option)
        };
        if distance <= threshold {
            result.push((option, distance));
        }
    }
    result.sort_by(|x, y| x.1.cmp(&y.1));
    result.into_iter().map(|(s, _)| s.to_string()).collect()
}

const MAX_SUGGESTIONS: usize = 5;

/// Given [ A, B, C ], returns "Did you mean A, B, or C?".
pub(crate) fn did_you_mean(suggestions: impl IntoIterator<Item = String>) -> String {
    const MESSAGE: &str = "Did you mean ";

    let suggestion_str = human_readable::join_strings(
        suggestions
            .into_iter()
            .take(MAX_SUGGESTIONS)
            .map(|s| format!("\"{}\"", s)),
        human_readable::JoinStringsOptions {
            separator: ", ",
            first_separator: None,
            last_separator: Some(", or "),
            output_length_limit: None,
        },
    );
    format!("{}{}?", MESSAGE, suggestion_str)
}
