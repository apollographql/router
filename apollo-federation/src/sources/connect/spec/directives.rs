use apollo_compiler::{
    ast::{Directive, Name, Value},
    schema::Component,
    Node, NodeStr,
};

use crate::error::FederationError;

use super::schema::{
    HTTPArguments, SourceDirectiveArguments, SOURCE_BASE_URL_ARGUMENT_NAME,
    SOURCE_HEADERS_ARGUMENT_NAME, SOURCE_HTTP_ARGUMENT_NAME, SOURCE_NAME_ARGUMENT_NAME,
};

impl TryFrom<&Component<Directive>> for SourceDirectiveArguments {
    type Error = FederationError;

    // TODO: This currently does not handle validation
    fn try_from(value: &Component<Directive>) -> Result<Self, Self::Error> {
        let ref args = value.arguments;

        // We'll have to iterate over the arg list and keep the properties by their name
        let mut name = None;
        let mut http = None;
        for arg in args {
            let arg_name = arg.name.as_str();

            if arg_name == SOURCE_NAME_ARGUMENT_NAME.as_str() {
                name = Some(
                    arg.value
                        .as_node_str()
                        .expect("`name` field in `@source` directive is not a string")
                        .clone(),
                );
            } else if arg_name == SOURCE_HTTP_ARGUMENT_NAME.as_str() {
                let http_value = arg
                    .value
                    .as_object()
                    .expect("`http` field in `@source` directive is not an object");
                let http_value = HTTPArguments::try_from(http_value)?;

                http = Some(http_value);
            }
        }

        // TODO: The compiler should catch missing fields here, right?
        Ok(Self {
            name: name.expect("missing `name` field in `@source` directive"),
            http: http.expect("missing `http` field in `@source` directive"),
        })
    }
}

impl TryFrom<&[(Name, Node<Value>)]> for HTTPArguments {
    type Error = FederationError;

    // TODO: This does not currently do validation
    fn try_from(values: &[(Name, Node<Value>)]) -> Result<Self, Self::Error> {
        // Iterate over all of the argument pairs and fill in what we need
        let mut base_url = None;
        let mut headers = None;
        for (name, value) in values {
            let name = name.as_str();

            if name == SOURCE_BASE_URL_ARGUMENT_NAME.as_str() {
                let base_url_value = value.as_node_str().expect(
                    "`baseURL` field in `@source` directive's `http` field is not a string",
                );

                base_url = Some(base_url_value.clone());
            } else if name == SOURCE_HEADERS_ARGUMENT_NAME.as_str() {
                let headers_value = value
                    .as_list()
                    .expect("`headers` field in `@source` directive's `http` field is not a list");

                headers = Some(todo!("handle headers for @source: {:?}", headers_value));
            }
        }

        Ok(Self {
            base_url: base_url
                .expect("missing `base_url` field in `@source` directive's `http` argument"),
            headers: headers.unwrap_or_default(),
        })
    }
}

#[cfg(test)]
mod tests {
    use apollo_compiler::Schema;

    use crate::{
        schema::ValidFederationSchema,
        sources::connect::spec::schema::{
            SourceDirectiveArguments, CONNECT_DIRECTIVE_NAME_IN_SPEC, SOURCE_DIRECTIVE_NAME_IN_SPEC,
        },
    };

    const SUBGRAPH_SCHEMA: &str = r#"
        extend schema
         @link(url: "https://specs.apollo.dev/connect/v0.1", import: ["@connect", "@source"])
         @source(
           name: "json"
           http: { baseURL: "https://jsonplaceholder.typicode.com/" }
         )

        type Query {
          users: [User]
           @connect(
             source: "json"
             http: { GET: "/users" }
             selection: "id name"
           )

          posts: [Post]
           @connect(
             source: "json"
             http: { GET: "/posts" }
             selection: "id title body"
           )
        }

        type User {
          id: ID!
          name: String
        }

        type Post {
          id: ID!
          title: String
          body: String
        }
    "#;

    #[test]
    fn it_parses_at_source() {
        let schema_str = format!(
            "{}\n{}\n{}",
            SUBGRAPH_SCHEMA, TEMP_FEDERATION_DEFINITIONS, TEMP_SOURCE_DEFINITIONS
        );
        let schema = Schema::parse(schema_str, "schema.graphql").unwrap();

        let schema = ValidFederationSchema::new(schema.validate().unwrap()).unwrap();

        let actual_definition = schema
            .get_directive_definition(&SOURCE_DIRECTIVE_NAME_IN_SPEC)
            .unwrap()
            .get(&schema.schema())
            .unwrap();

        insta::assert_snapshot!(
            actual_definition.to_string(),
            @r###"
                """
                Defines connector configuration for reuse across multiple connectors.

                Exactly one of {http} must be present.
                """
                directive @source(name: String!, http: SourceHTTP) on SCHEMA
            "###
        );

        insta::assert_debug_snapshot!(
            schema
                .referencers()
                .get_directive(SOURCE_DIRECTIVE_NAME_IN_SPEC.as_str())
                .unwrap(),
            @r###"
                DirectiveReferencers {
                    schema: Some(
                        SchemaDefinitionPosition,
                    ),
                    scalar_types: {},
                    object_types: {},
                    object_fields: {},
                    object_field_arguments: {},
                    interface_types: {},
                    interface_fields: {},
                    interface_field_arguments: {},
                    union_types: {},
                    enum_types: {},
                    enum_values: {},
                    input_object_types: {},
                    input_object_fields: {},
                    directive_arguments: {},
                }
            "###
        );
    }

    #[test]
    fn it_parses_at_connect() {
        let schema_str = format!(
            "{}\n{}\n{}",
            SUBGRAPH_SCHEMA, TEMP_FEDERATION_DEFINITIONS, TEMP_SOURCE_DEFINITIONS
        );
        let schema = Schema::parse(schema_str, "schema.graphql").unwrap();

        let schema = ValidFederationSchema::new(schema.validate().unwrap()).unwrap();

        let actual_definition = schema
            .get_directive_definition(&CONNECT_DIRECTIVE_NAME_IN_SPEC)
            .unwrap()
            .get(&schema.schema())
            .unwrap();

        insta::assert_snapshot!(
            actual_definition.to_string(),
            @r###"
              """
              Defines a connector as the implementation of a field.

              Exactly one of {http} must be present.
              """
              directive @connect(
                """
                Optionally connects a @source directive for shared connector configuration.
                Must match the `name:` argument of a @source directive in this schema.
                """
                source: String,
                """Defines HTTP configuration for this connector."""
                http: ConnectHTTP,
                """
                Uses the JSONSelection syntax to define a mapping of connector response
                to GraphQL schema.
                """
                selection: JSONSelection,
                """
                Marks this connector as a canonical resolver for an entity (uniquely
                identified domain model.) If true, the connector must be defined on a
                field of the Query type.
                """
                entity: Boolean = false,
              ) on FIELD_DEFINITION
        "###
        );

        let fields = schema
            .referencers()
            .get_directive(CONNECT_DIRECTIVE_NAME_IN_SPEC.as_str())
            .unwrap()
            .object_fields
            .iter()
            .map(|f| f.get(&schema.schema()).unwrap().to_string())
            .collect::<Vec<_>>()
            .join("\n");

        insta::assert_snapshot!(
            fields,
            @r###"
                users: [User] @connect(source: "json", http: {GET: "/users"}, selection: "id name")
                posts: [Post] @connect(source: "json", http: {GET: "/posts"}, selection: "id title body")
            "###
        );
    }

    #[test]
    fn it_extracts_at_source() {
        // Convert the schema into its validated form
        let schema_str = format!(
            "{}\n{}\n{}",
            SUBGRAPH_SCHEMA, TEMP_FEDERATION_DEFINITIONS, TEMP_SOURCE_DEFINITIONS
        );
        let schema = Schema::parse(schema_str, "schema.graphql").unwrap();
        let schema = ValidFederationSchema::new(schema.validate().unwrap()).unwrap();

        // Try to extract the source information from the valid schema
        // TODO: This should probably be handled by the rest of the stack
        let sources = schema
            .referencers()
            .get_directive(&SOURCE_DIRECTIVE_NAME_IN_SPEC)
            .unwrap();

        // Extract the sources from the schema definition and map them to their `Source` equivalent
        // TODO: We can safely assume that a source can only be on a schema, right?
        let schema_directive_refs = sources.schema.as_ref().unwrap();
        let sources: Result<Vec<_>, _> = schema_directive_refs
            .get(schema.schema())
            .directives
            .iter()
            .filter(|directive| directive.name == SOURCE_DIRECTIVE_NAME_IN_SPEC)
            .map(SourceDirectiveArguments::try_from)
            .collect();

        insta::assert_debug_snapshot!(
            sources.unwrap(),
            @r###"
                [
                    SourceDirectiveArguments {
                        name: "json",
                        http: HTTPArguments {
                            base_url: "https://jsonplaceholder.typicode.com/",
                            headers: [],
                        },
                    },
                ]
            "###
        );
    }

    static TEMP_FEDERATION_DEFINITIONS: &str = r#"
        directive @link(url: String, as: String, for: link__Purpose, import: [link__Import]) repeatable on SCHEMA
        scalar link__Import

        enum link__Purpose {
          """
          \`SECURITY\` features provide metadata necessary to securely resolve fields.
          """
          SECURITY

          """
          \`EXECUTION\` features provide metadata necessary for operation execution.
          """
          EXECUTION
        }
    "#;

    static TEMP_SOURCE_DEFINITIONS: &str = r#"
        """
        Defines a connector as the implementation of a field.

        Exactly one of {http} must be present.
        """
        directive @connect(
          """
          Optionally connects a @source directive for shared connector configuration.
          Must match the `name:` argument of a @source directive in this schema.
          """
          source: String

          """
          Defines HTTP configuration for this connector.
          """
          http: ConnectHTTP

          """
          Uses the JSONSelection syntax to define a mapping of connector response
          to GraphQL schema.
          """
          selection: JSONSelection

          """
          Marks this connector as a canonical resolver for an entity (uniquely
          identified domain model.) If true, the connector must be defined on a
          field of the Query type.
          """
          entity: Boolean = false
        ) on FIELD_DEFINITION

        """
        HTTP configuration for a connector.

        Exactly one of {GET,POST,PATCH,PUT,DELETE} must be present.
        """
        input ConnectHTTP {
          """
          URL template for GET requests to an HTTP endpoint.

          Can be a full URL or a partial path. If it's a partial path, it will
          be appended to an associated `baseURL` from the related @source.
          """
          GET: URLPathTemplate

          "Same as GET but for POST requests"
          POST: URLPathTemplate

          "Same as GET but for PATCH requests"
          PATCH: URLPathTemplate

          "Same as GET but for PUT requests"
          PUT: URLPathTemplate

          "Same as GET but for DELETE requests"
          DELETE: URLPathTemplate

          """
          Define a request body using JSONSelection. Selections can include
          values from field arguments using `$args.argName` and from fields on the
          parent type using `$this.fieldName`.
          """
          body: JSONSelection

          """
          Configuration for headers to attach to the request.

          Takes precedence over headers defined on the associated @source.
          """
          headers: [HTTPHeaderMapping!]
        }

        """
        At most one of {as,value} can be present.
        """
        input HTTPHeaderMapping {
          "The name of the incoming HTTP header to propagate to the endpoint"
          name: String!

          "If present, this defines the name of the header in the endpoint request"
          as: String

          "If present, this defines values for the headers in the endpoint request"
          value: [String]
        }

        """
        Defines connector configuration for reuse across multiple connectors.

        Exactly one of {http} must be present.
        """
        directive @source(
          name: String!

          http: SourceHTTP
        ) on SCHEMA

        """
        Common HTTP configuration for connectors.
        """
        input SourceHTTP {
          """
          If the URL path template in a connector is not a valid URL, it will be appended
          to this URL. Must be a valid URL.
          """
          baseURL: String!

          """
          Common headers from related connectors.
          """
          headers: [HTTPHeaderMapping!]
        }

        """
        A string containing a "JSON Selection", which defines a mapping from one JSON-like
        shape to another JSON-like shape.

        Example: ".data { id: user_id name account: { id: account_id } }"
        """
        scalar JSONSelection @specifiedBy(url: "...")

        """
        A string that declares a URL path with values interpolated inside `{}`.

        Example: "/product/{$this.id}/reviews?count={$args.count}"
        """
        scalar URLPathTemplate @specifiedBy(url: "...")
    "#;
}
