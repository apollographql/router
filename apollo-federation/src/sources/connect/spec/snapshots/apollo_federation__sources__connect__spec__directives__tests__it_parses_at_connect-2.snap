---
source: src/sources/connect/spec/directives.rs
expression: fields
---
users: [User] @connect(source: "json", http: {GET: "/users"}, selection: "id name")
posts: [Post] @connect(source: "json", http: {GET: "/posts"}, selection: "id title body")
